theano/gpuarray/reduction.py

```diff
@@ -110,80 +81,68 @@ def c_code(self, node, name, input_names, output_names, sub):
         # Recall: fail = sub['fail']
         max_typecode = pygpu.gpuarray.dtype_to_typecode(node.inputs[0].dtype)
         argmax_typecode = pygpu.gpuarray.dtype_to_typecode(self.argmax_dtype)
-        # axes_ctype = pygpu.gpuarray.dtype_to_ctype(node.inputs[1].dtype)
         axes_ctype = 'int64_t'
+        assert node.inputs[1].ndim == 1
         ret = """
+        GpuArray temp;
         GpuArray* %(name)s_input = &%(X)s->ga;
         size_t %(name)s_input_ndim = PyGpuArray_NDIM(%(X)s);
-        """
-        if NoneConst.equals(node.inputs[1]):
-            ret += """
-            unsigned  %(name)s_redux_len = %(name)s_input_ndim;
-            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
-            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
-                %(name)s_axes_to_reduce[i] = i;
-            }
-            """
-        else:
-            assert node.inputs[1].ndim == 1
-            ret += """
-            unsigned  %(name)s_redux_len = PyArray_DIM(%(axes)s, 0);
-            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
-            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
-                %(name)s_axes_to_reduce[i] =
-                    (unsigned)( ((%(axes_ctype)s*)PyArray_DATA(%(axes)s)) [i * (PyArray_STRIDES(%(axes)s)[0] / sizeof(%(axes_ctype)s))] );
-            }
-            """
-        ret += """
+
+        unsigned  %(name)s_redux_len = PyArray_DIM(%(axes)s, 0);
+        unsigned* %(name)s_axes_to_reduce = (unsigned*)malloc(%(name)s_redux_len * sizeof(unsigned));
+        for (unsigned i = 0; i < %(name)s_redux_len; ++i) {
+            %(name)s_axes_to_reduce[i] = (unsigned) (*(%(axes_ctype)s*)PyArray_GETPTR1(%(axes)s, i));
+        }
+
         size_t  %(name)s_output_ndim = %(name)s_input_ndim - %(name)s_redux_len;
         size_t* %(name)s_output_dims = NULL;
-        if(%(name)s_output_ndim == 0) {
+        if (%(name)s_output_ndim == 0) {
             /* Current backend function GpuArray_maxandargmax does not work when
              * all axes need to be reduced. So to handle this case, we create a view
              * of the input as a matrix with 1 row and as many columns as elements
              * in the input, so that the 2nd dimenson of the matrix will be reduced. */
```

Comment:
Yes ! Please see my previous comment about it !

When all axes need to be reduced, I must reshape, or the execution fails with a runtime error:

```
RuntimeError: Unexpected error in an Op's C code. No Python exception was set.
Apply node that caused the error: GpuMaxAndArgmax(<GpuArrayType<None>(float32, (False, False, False, False, False))>, TensorConstant{[0 1 2 3 4]})
Toposort index: 0
Inputs types: [GpuArrayType<None>(float32, (False, False, False, False, False)), TensorType(int64, vector)]
Inputs shapes: [(1000, 100, 10, 5, 2), (5,)]
Inputs strides: [(40000, 400, 40, 8, 4), (8,)]
Inputs values: ['not shown', array([0, 1, 2, 3, 4])]
Outputs clients: [[HostFromGpu(gpuarray)(GpuMaxAndArgmax.0)], [HostFromGpu(gpuarray)(GpuMaxAndArgmax.1)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```

It seems it is related to libgpuarray.